param/parameterized.py

```diff
@@ -403,19 +403,14 @@ def iscoroutinefunction(function):
 def _instantiate_param_obj(paramobj, owner=None):
     """Return a Parameter object suitable for instantiation given the class's Parameter object"""
 
-    # Shallow-copy Parameter object
-    p = copy.copy(paramobj)
+    # Shallow-copy Parameter object without the watchers
+    try:
+        watchers = paramobj.watchers
+        paramobj.watchers = {}
+        p = copy.copy(paramobj)
+    finally:
+        paramobj.watchers = watchers
     p.owner = owner
-
-    # Reset watchers since class parameter watcher should not execute
-    # on instance parameters
-    p.watchers = {}
-
-    # shallow-copy any mutable slot values other than the actual default
-    for s in p.__class__.__slots__:
```

Comment:
It seemed a little silly to set slot values twice (once in `__setstate__` and once here), but I think I should revert this.